File: cartography/feed.go
package cartography

import (
	"context"
	"sort"
	"sync"
	"time"
)

// DataSignal is a single quantitative input — one FRED series reduced to a
// value, a threshold, and a triggered/clear status the algorithm can act on.
type DataSignal struct {
	Source      string    `json:"source"`      // e.g. "FRED:UNRATE"
	Name        string    `json:"name"`        // human-readable
	Value       float64   `json:"value"`       // current observation
	Threshold   float64   `json:"threshold"`   // trigger level
	Triggered   bool      `json:"triggered"`   // is the signal hot?
	AsOf        time.Time `json:"as_of"`       // observation date
	Description string    `json:"description"` // what the value means
}

// DataFeed is the composed real-data picture: each individual signal plus
// the multiplier they jointly imply.
type DataFeed struct {
	Signals     []DataSignal `json:"signals"`
	Triggers    []string     `json:"triggers"`     // names of fired signals
	Multiplier  float64      `json:"multiplier"`   // composed risk scalar
	Warnings    []string     `json:"warnings"`     // partial-fetch problems
	LastFetched time.Time    `json:"last_fetched"` // cache timestamp
}

// signalHaircut maps a triggered signal to the multiplicative haircut it
// applies to the position-size scalar. Intentionally conservative — the
// purpose is to *cap* risk when reality is hostile, not to chase.
//
// Sahm 0.40        — historically near-perfect recession trigger
// HY stress 0.50   — credit dislocation always precedes equity drawdowns
// Yield inv 0.85   — leading, not coincident; light haircut
// NFCI tight 0.85  — tightening conditions; light haircut
//
// Two small haircuts compound (0.85 × 0.85 ≈ 0.72), which matches how
// these signals tend to fire in clusters before something breaks.
func signalHaircut(name string) float64 {
	switch name {
	case "Sahm Recession Indicator":
		return 0.40
	case "High-Yield Credit Spread (OAS)":
		return 0.50
	case "10Y-2Y Treasury Spread":
		return 0.85
	case "Financial Conditions (NFCI)":
		return 0.85
	default:
		return 1.0
	}
}

// composeMultiplier reduces the active signals to a single risk scalar by
// multiplying their haircuts together. Returns (multiplier, names of
// triggered signals).
func composeMultiplier(signals []DataSignal) (float64, []string) {
	mult := 1.0
	var fired []string
	for _, s := range signals {
		if s.Triggered {
			mult *= signalHaircut(s.Name)
			fired = append(fired, s.Name)
		}
	}
	sort.Strings(fired)
	return mult, fired
}

// FeedCache keeps a cached snapshot of the FRED feed and refreshes it on
// a schedule rather than on every request — FRED has rate limits and the
// underlying data only updates daily.
type FeedCache struct {
	client  *FREDClient
	mu      sync.RWMutex
	current *DataFeed
	maxAge  time.Duration
}

// NewFeedCache builds a cache. maxAge of 0 defaults to 6 hours.
func NewFeedCache(client *FREDClient, maxAge time.Duration) *FeedCache {
	if maxAge <= 0 {
		maxAge = 6 * time.Hour
	}
	return &FeedCache{client: client, maxAge: maxAge}
}

// Get returns the cached feed (may be nil if never fetched). It does not
// trigger a refresh — the caller schedules that separately.
func (c *FeedCache) Get() *DataFeed {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.current
}

// Refresh pulls a fresh snapshot from FRED and stores it. Safe to call
// concurrently — the last writer wins. Returns the new feed plus any
// fetch error (the cache is not updated on hard error).
func (c *FeedCache) Refresh(ctx context.Context) (*DataFeed, error) {
	feed, err := c.client.Snapshot(ctx)
	if err != nil {
		return nil, err
	}
	c.mu.Lock()
	c.current = feed
	c.mu.Unlock()
	return feed, nil
}

// AppliedMultiplier picks the more cautious of the formula multiplier and
// the data multiplier. The formula provides a slow-moving prior; the feed
// provides a coincident veto. min() means the feed can shrink risk when
// data disagrees with the model, but cannot enlarge it past what the
// formula already allows.
func AppliedMultiplier(formula float64, feed *DataFeed) float64 {
	if feed == nil {
		return formula
	}
	if feed.Multiplier < formula {
		return feed.Multiplier
	}
	return formula
}

File: rust-trader/crates/cartography/src/feed.rs
//! Feed types, cache, and multiplier composition.

use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use std::sync::{Arc, RwLock};

use crate::fred::FREDClient;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataSignal {
    pub source: String,
    pub name: String,
    pub value: f64,
    pub threshold: f64,
    pub triggered: bool,
    pub as_of: DateTime<Utc>,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataFeed {
    pub signals: Vec<DataSignal>,
    pub triggers: Vec<String>,
    pub multiplier: f64,
    pub warnings: Vec<String>,
    pub last_fetched: DateTime<Utc>,
}

fn signal_haircut(name: &str) -> f64 {
    match name {
        "Sahm Recession Indicator" => 0.40,
        "High-Yield Credit Spread (OAS)" => 0.50,
        "Investment-Grade Credit Spread (OAS)" => 0.75,
        "10Y-2Y Treasury Spread" => 0.85,
        "Financial Conditions (NFCI)" => 0.85,
        _ => 1.0,
    }
}

pub fn compose_multiplier(signals: &[DataSignal]) -> (f64, Vec<String>) {
    let mut mult = 1.0;
    let mut fired = Vec::new();
    for s in signals {
        if s.triggered {
            mult *= signal_haircut(&s.name);
            fired.push(s.name.clone());
        }
    }
    fired.sort();
    (mult, fired)
}

pub fn applied_multiplier(formula: f64, feed: Option<&DataFeed>) -> f64 {
    match feed {
        Some(f) if f.multiplier < formula => f.multiplier,
        _ => formula,
    }
}

pub struct FeedCache {
    #[allow(dead_code)]
    client: FREDClient,
    current: Arc<RwLock<Option<DataFeed>>>,
    #[allow(dead_code)]
    max_age: Duration,
}

impl FeedCache {
    pub fn new(client: FREDClient, max_age: Duration) -> Self {
        Self {
            client,
            current: Arc::new(RwLock::new(None)),
            max_age: if max_age.num_seconds() == 0 {
                Duration::hours(6)
            } else {
                max_age
            },
        }
    }

    pub fn get(&self) -> Option<DataFeed> {
        self.current.read().ok()?.clone()
    }

    pub fn get_fresh(&self, now: DateTime<Utc>) -> Option<DataFeed> {
        let feed = self.get()?;
        if now.signed_duration_since(feed.last_fetched) <= self.max_age {
            Some(feed)
        } else {
            None
        }
    }

    pub async fn refresh(&self) -> Result<DataFeed, Box<dyn std::error::Error + Send + Sync>> {
        let feed = self.client.snapshot().await?;
        *self.current.write().unwrap() = Some(feed.clone());
        Ok(feed)
    }
}
